### DISCUSSION TITLE
OpenTelemetry log exporter and ElasticSearch APM 8+

### DISCUSSION THREAD

**Discussion by p10tyr** [open-telemetry/opentelemetry-dotnet#3249] [Q&A] [answered]
Hi

I am just trying to see if I can get logs into Elastic Stack. It seems that since ES version 8 it is supported with the unstable data model for now. (I am on ES 8.1)

Tracking this is seems like there should be some very basic stuff coming though by now
https://github.com/open-telemetry/opentelemetry-dotnet/issues/2482

I am looking at this example https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/examples/Console/TestLogs.cs and it seems to use a local collector but I do not see where the URI is being configured? Is the default just localhost:4318 or something as per the docker instructions? Just curios because I am setting my own endpoint and not sure what the default value would be?? if any

So for my logging configuration I am using as below and I don't think I need to explicitly specify GRPC (but that is the only protocol Elastics supports for now) neither the simple exporter?

Sadly I do not see anything show up in Elastic streams and by the looks of https://github.com/open-telemetry/opentelemetry-dotnet/issues/2482 there should be some basic data coming through 

Using OpenTelemetry.Exporter.OpenTelemetryProtocol.Logs 1.0.0-rc9.2 (lowest to match)
To match stable OpenTelemetry 1.2.0

Unless I need to bump up to Otel 1.3.0 and logs-rc9.3? (I tried but because some of our stuff still used 4.7.2 I cant bump up other dependencies without a massive fallout) 

```
 .AddLogging(configure => configure.AddOpenTelemetry(options =>
                {
                    options.IncludeScopes = true;

                    options.AddOtlpExporter(otlpExporter =>
                    {
                        otlpExporter.Protocol = OpenTelemetry.Exporter.OtlpExportProtocol.Grpc;
                        otlpExporter.ExportProcessorType = ExportProcessorType.Simple;
                        otlpExporter.Endpoint = new System.Uri("https://xyz.apm.aws.cloud.es.io");
                        otlpExporter.Headers = "Authorization=Bearer supersecret";
                    });
                }));
```

**Comment by cijothomas** [comment, marked as answer]
Please take the newest version from myget - https://github.com/open-telemetry/opentelemetry-dotnet#release-schedule
(as some of the OTLPLogExporter fixes are very new and hasn't released to nuget yet).

Or please collect sdk logs https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry#self-diagnostics that can tell if something is wrong within the sdk/exporter.

**Reply by p10tyr** [reply]
I have cloned the latest main of the repository and build the Examples.Console.

I wired in the ElasticSearch APM URL just to see if it works but nothing shows up. I've run it several times over 15 minutes.

So I followed the instructions on to enable self diagnosis but there is no log file being created. I also dropped the level to logalways.. still no file being generated.

I run it via Visual studio and by just running the exe from the debug folder in terminal.

So I enabled all exceptions for CLR Exceptions to see if anything is not happy.. since I have all the source code it should be easy for me to try and find some issues.?

So only thing I get is A task was cancelled in .. (because the console has basically started to dispose after doing the TestLog Logger stuff)

SelfDiagnosticsConfigRefresher.Task Worker(CancellationToken cancellationToken)

And that was about it.. Nothing else errored out and the Task Cancelled is pretty normal behavior? 

I am going to try and get the Open Telemetry Docker running now... and point to that see if I get any outputs.

If you have any more ideas please let me know

**Reply by p10tyr** [reply]
So I changed the TestLogs to be like 
```
 opt.AddOtlpExporter(otlpOptions =>
                    {
                        otlpOptions.Protocol = OpenTelemetry.Exporter.OtlpExportProtocol.Grpc;
                        otlpOptions.Endpoint = new Uri("http://127.0.0.1:4317");
                    });
```
And it does infact emit OTLP Logs over gRPC. 

![image](https://user-images.githubusercontent.com/1249683/166816717-acc2aae5-fa1e-4acc-8eb8-4329d4dd47e3.png)

So I guess.. Elastics APM Agent is ignoring everything 😭

But at least I know how to debug OTLP now. I have the source code building too and I am still going to try and figure out what is going on with ES

I have hooked in my real world application with the version I mentioned before and the logs and traces with the minimum data I need are showing up .

![image](https://user-images.githubusercontent.com/1249683/166818305-7b7785d8-6cb1-44e9-82d8-c3331d4b8b00.png)

**Reply by cijothomas** [reply]
> So I guess.. Elastics APM Agent is ignoring everything 😭

If it was rejecting the data, I think the Exporter should have logged an exception which would have shown up in the Self diag logs

**Reply by p10tyr** [reply]
I am basically following this (just in C#) https://github.com/elastic/apm-server/pull/6768

Although they spin up a collector.. I wonder if they want us to run an agent like the old days- its unclear to me. The release notes says the cloud agent supports it .. im looking for logs in ES

**Reply by cijothomas** [reply]
Once your figure out the answer, please post back what you learned. 
I never tried ES with OTLP, so unsure if it can natively ingest OTLP....

**Reply by p10tyr** [reply]
I sure will do. But just so you know..

> Elastic APM Server natively supports the OpenTelemetry protocol. 

https://www.elastic.co/guide/en/apm/guide/current/open-telemetry.html

And FYI:
All the tracing and metrics works great.

**Reply by p10tyr** [reply]
Well here is a revelation! (using TestLogs)

It now decided to show up in log stream (maybe there is a long delay in processing)
Problem is I changed stuff so many times I don't know what the combination was

![image](https://user-images.githubusercontent.com/1249683/166822128-764c1dd4-1bca-4a81-8ce9-83912e5aa900.png)

**Reply by cijothomas** [reply]
![image](https://user-images.githubusercontent.com/5232798/166822358-e207e257-bba3-4214-91ab-b9b163d9dcdc.png)
It looks like https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter has a elasticsearchexporter and elasticexporter, the latter being not recommended due to native support added...

**Reply by p10tyr** [reply]
Yea. I have no idea. After spending the whole day today no logs were showing up from my real application. 

Now, all of a sudden. No code changes .. its streaming.

So to confirm I am using OpenTelemetry 1.2.0 with OpenTelemetry.Exporter.OpenTelemetryProtocol.Logs 1.0.0-rc9.2

I know some things are missing but for now just the logs and correlation ID's are super helpful.

I just wish I knew what it was that started the Logs to stream in Elastics.. I must have gone somewhere or clicked something while messing around in the Management settings that kicked it off ? Very confused.

Thank you for your help!

**Reply by cijothomas** [reply]
Please use the newest versions as they release (or from myget feed..), and share feedbacks. OTel Logging SIG is about to declare Logs Protocol as stable, so we are unblocked from that perspective to release a Stable version of OTLP LogExporter.

**Reply by p10tyr** [reply]
My real world application I cant just use your package because it is trying to upgrade a lot of Net packages.. that I cant just do with out a massive headache :(  I am dealing with a hybrid / half migrated application.. So I will do that when I have 1 week to burn. 

But this is good new regardless. I cant wait for it to go stable!

I would like ES Exporter to be thing.. Wonder if I should take a look at writing something because I am adding extra attributes to enrich ES logs. I will provide them here for the use of other people.

```
var otelAttributes = new List<KeyValuePair<string, object>>();
              otelAttributes.Add(new KeyValuePair<string, object>("deployment.environment", environmentName));

options.SetResourceBuilder(ResourceBuilder.CreateDefault()
   .AddService("ServiceName")  <-- Populates the service name if you need it
   .AddAttributes(otelAttributes));  <-- Enables filtering by environment in ES (also add any extra attributes if you need them)
```

**Comment by ejsmith** [comment]
I just got everything being sent to Elasticsearch 8.2 using otel collector and the ES apm server. It was a huge PITA to get it all working. You can view my setup in this branch: https://github.com/exceptionless/Exceptionless/tree/feature/elastic8x

Docker compose to run ES 8.x, Kibana and APM server all talking to each other with SSL turned on because it's required for APM server.
https://github.com/exceptionless/Exceptionless/blob/feature/elastic8x/docker-compose.yml

Otel collector config for collecting and forwarding to APM server:
https://github.com/exceptionless/Exceptionless/blob/feature/elastic8x/otel-config.yml

Ideally, we should be able to send directly to the APM Server and not have to run an otel collector instance, but I wasn't able to get it working.

My .NET configuration code to collect traces, metrics and logs from my app:
https://github.com/exceptionless/Exceptionless/blob/feature/elastic8x/src/Exceptionless.Web/ApmExtensions.cs

**Reply by p10tyr** [reply]
I am sending directly to Elastic Cloud APM Agent that is running in Fleet. Version 8.1

```
 otlpExporter.Endpoint = new Uri("https://redacted.apm.redacted.aws.cloud.es.io");
 otlpExporter.Headers = "Authorization=Bearer redacted";
```

![image](https://user-images.githubusercontent.com/1249683/166832575-91d0800a-0c4f-46a1-a5f8-d67b913e0320.png)

To be fair. I am not 100% certain how it is all configured in the background. Maybe they spin up a collector on our behalf and that is why I was not seeing any thing before.

Maybe I triggered something when I ran this repo test code and all of a sudden logs started spitting out even for my older versions

**Reply by p10tyr** [reply]
Really cool stuff in the ApmExtensions! I am going to be going over that tomorrow morning 💯 ❤️

**Reply by ejsmith** [reply]
Yeah, it's a LOT easier that way. :laughing:

**Reply by p10tyr** [reply]
I'm definitely going to give your service a test drive with another project I'm working on when I sort this out 😄 
Can't say that I have seen or heard of this before. Looks really awesome.

**Reply by ejsmith** [reply]
Awesome! Let me know if you need anything or have any questions.

**Reply by Shailen31** [reply]
Hey @ejsmith  

in response to : "Docker compose to run ES 8.x, Kibana and APM server all talking to each other with SSL turned on because it's required for APM server."


not sure if this has changed but this is a excerpt from the documentation for v8.17 
"APM Server does not require agents to provide a certificate for authentication, and there is no dedicated support for SSL/TLS client certificate authentication in Elastic’s backend agents." - [in the docs](https://www.elastic.co/guide/en/observability/current/apm-agent-tls.html#:~:text=TLS%20is%20disabled%20by%20default,corresponding%20private%20key%20are%20required.)

So from your experience, can you send telemetry data to a local apm server running in docker without any tls enabled?
ive been struggling to get this working as i keep getting the error: "rpc error: code = unavailable desc = connection error: desc = \"transport: authentication handshake failed: tls: first record does not look like a tls handshake\"", "interval": "23.210157887s"}"

but the documentation seems to give a conflicting statement unless i am missing something or conflating 

thanks
